### DISCUSSION TITLE
Enable gunicorn preloading to reduce RAM usage?

### DISCUSSION THREAD

**Discussion by stapelberg** [pretix/pretix#2386] [Ideas] [answered]
I was surprised to discover that starting pretix on my server reduces the available memory by about 1 GB!

It looks like the 4 gunicorn worker processes are the most expensive, with each consuming 8% RSS (158 MB on this VM with 2G RAM).

I found https://docs.gunicorn.org/en/stable/settings.html#preload-app and was wondering if the `--preload` setting is safe to enable with pretix, and if yes, how to best enable it? IIUC, enabling preloading should move most of the memory usage to the parent gunicorn process, with workers only slowly accumulating memory over time (which in turn can be curbed by decreasing `--max-requests` and/or decreasing the number of workers).

AFAICT, gunicorn is started at https://github.com/pretix/pretix/blob/c2a8286022b58f872173d8e1b1d62ffc71fbf356/deployment/docker/pretix.bash#L35, and I don’t see a way to add the `--preload` flag easily.

I’m using the Docker small-scale deployment in case it matters.

**Comment by raphaelm** [comment, marked as answer]
Let's start with the short answers first:

### How to best enable it?

In the docker setup, currently only with a custom Dockerfile that replaces pretix.bash.

### Is it safe?

I don't know. I hope it is. Django's design patterns discourage use of global state, and in there rare occasions we do need global state, both Django itself as well as our own code use ``threading.local`` for proper isolation, so I'd assume our own code as well as Django itself are mostly safe. However, I have not audited all of our dependencies for use of global state and for other import-time side effects. I'd be at least a little careful and wouldn't feel great deploying this into production without finding a good way to stress-test it first.

### Is it worthwile?

I don't know, it would be interesting to measure the actual benefit with real-world load.

Obviously using less RAM would be great. But I'm skeptical for the following reasons:

- However, as you already mentioned, this will cause a much more load-dependent use of RAM. Currently, the RAM use of our gunicorn workers is pretty much constant, even under high load. Unless you add more workers, the 1GB RAM won't significantly increase¹. Since ticketing is a use case that can typically produce unexpected peak loads, it is quite useful to know that one only needs to worry about having enough CPU capacity (and a fast database) to withstand high load, instead of needing to monitor and worry about about multiple system parameters at once. As you already mentioned, ``max-requests`` helps with that, but booting a new worker is very costly² and reducing ``max-requests`` significantly will lead to many many worker restarts *especially* under high load, leading to *even more* pressure on the CPU, which is usually our bottleneck already.

- "Import before forking" is generally considered pretty useless for pure Python code, due to CPython's reference counting mechanism [causing copy on write](https://bugs.python.org/issue40255) early and often. This [Instagram blog post](https://instagram-engineering.com/dismissing-python-garbage-collection-at-instagram-4dca40b29172) gives some numbers (shared memory drops down again very quickly after fork), and also indicates that refcounting might not be the actual issue, but the garbage collector is. Instagram ended up completely disabling garbage collection, but as the blog post shows this creates new risks as well.

- As stated above, running pretix under load is very much CPU-bound. Having a quick look at [different](https://www.hetzner.com/cloud) [VPS](https://www.ovhcloud.com/de/vps/) and [cloud](https://aws.amazon.com/de/ec2/instance-types/) providers, today's general purpose instances or compute-optimized instances are mostly specced with 2-4 GB per CPU core. For bare metal machines, the ratio is even higher. Therefore, when deciding for the proper hosting environment for a pretix installation, we usually already only need to look at the number of CPUs, since there will be "enough RAM" by default. Reducing RAM footprint would therefore mostly benefit environments where resources are directly shared with other applications, which would still be nice, but don't make it a primary area of optimization efforts.

¹ This is true for our gunicorn workers, but not for the celery workers who obviously can have increased RAM usage if you e.g. use them to e.g. compile an export including hundreds of megabytes of data.
² At least currently, due to the many imports. ``--preload`` could in fact soften that a little.

**Reply by stapelberg** [reply]
Thanks for the detailed answer! I’ll not pursue this idea for now due to the reasons you described.
